clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -486,6 +508,20 @@ synchronized void adjustSequencesDueToFailedBatch(ProducerBatch batch) {
         }
     }
 
+    private synchronized void startSequencesAtBeginning(TopicPartition topicPartition) {
```

Comment:
Discussed offline. There are some edge cases where resetting the sequence number may not be safe. In particular, since retention policies are enforced asynchronously on the brokers, it could be possible to see an UnknownProducer exception on one broker and then a leader failover could effectively restore some of producer state. This case is rare, but maybe the safest way to handle it is to reset the producerId.

Reply:
The one offshoot of this is that _if_ we reset the producer id here, then the user is certain to hit https://issues.apache.org/jira/browse/KAFKA-5870, which will result in `OutOfOrderSequencExceptions` for all in flight requests. 

I wonder if that inconvenience is worth it. I am on the fence at this point, but leaning slightly toward leaving the code as is until we fix KAFKA-5870.